### PR TITLE
feat: [CDS-59358]:  Logic for cleaning failing V2 InstanceSyncPerpetaulTaskInfo

### PR DESCRIPTION
**Pull request by namantalaycha10**
## Describe your changes
logic to clear InstanceSyncPerpetaulTaskInfo if perpetual task execution fails for more than 7 days

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49221)

### CODE REVIEW

**Comment by namantalaycha10** [comment]
trigger uts

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 1]
```diff
@@ -115,6 +117,8 @@ public class InstanceSyncServiceImpl implements InstanceSyncService {
   private static final int RELEASE_COUNT_LIMIT =
       Integer.parseInt(System.getenv().getOrDefault("INSTANCE_SYNC_RESPONSE_BATCH_RELEASE_COUNT", "5"));
 
+  @VisibleForTesting static final long RELEASE_PRESERVE_TIME = TimeUnit.DAYS.toMillis(7);
```
Comment (abhinavsingh1196) [inline 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:120]: this annotation is needed? looks unused

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 2]
```diff
@@ -533,6 +542,15 @@ private void migrateToInstanceSyncV1(InstanceSyncPerpetualTaskInfoDTO instanceSy
     }
   }
 
+  private void cleanupFailedPerpetualTask(InstanceSyncPerpetualTaskInfoDTO instanceSyncPerpetualTaskInfoDTO) {
```
Comment (abhinavsingh1196) [inline 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:545]: should rename this method to something like `cleanPerpetualTaskIfFailingForSevenDays()`

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 3]
```diff
@@ -533,6 +542,15 @@ private void migrateToInstanceSyncV1(InstanceSyncPerpetualTaskInfoDTO instanceSy
...
+    if (instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() != 0
+        && instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() + RELEASE_PRESERVE_TIME
```
Comment (abhinavsingh1196) [inline 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:547]: isn't this check redundant ? `instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() != 0`
If it is equal to 0 by default, then the next condition would be false anyway

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java (hunk 3 above)
Comment (namantalaycha10) [reply 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:547]: no , if instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() is equal to 0
then this 
```
instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() + RELEASE_PRESERVE_TIME
            <  System.currentTimeMillis()
```
will be true
we are checking here less than condition

**Review by abhinavsingh1196 — APPROVED** [review]

**Comment by namantalaycha10** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Comment by namantalaycha10** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 4]
```diff
@@ -546,7 +546,8 @@ private void migrateToInstanceSyncV1(InstanceSyncPerpetualTaskInfoDTO instanceSy
 
   private void cleanPerpetualTaskIfFailingForSevenDays(
       InstanceSyncPerpetualTaskInfoDTO instanceSyncPerpetualTaskInfoDTO) {
-    if (instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() != 0
+    if (instanceSyncPerpetualTaskInfoDTO.getLastSuccessfulRun() != null
```
[the remark this answers is not in the document]
Comment (namantalaycha10) [reply 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:549]: Thanks @abhinavsingh1196 , sure

**Pull request merged**